src/pd/client.rs

```diff
@@ -34,16 +35,92 @@ const SOCKET_WRITE_TIMEOUT: u64 = 3;
 
 const PD_RPC_PREFIX: &'static str = "/pd/rpc";
 
+// Only for `validate_endpoints`.
+const VALIDATE_MSG_ID: u64 = 0;
+const VALIDATE_CLUSTER_ID: u64 = 0;
+
+// `validate_endpoints` validates pd members, make sure they are in the same cluster.
+// It returns a cluster ID.
+// Notice that it ignores failed pd nodes.
+fn validate_endpoints(endpoints: &[String]) -> Result<u64> {
+    if endpoints.is_empty() {
+        return Err(box_err!("empty PD endpoints"));
+    }
+
+    let len = endpoints.len();
+    let mut endpoints_set = HashSet::with_capacity(len);
+
+    let mut cluster_id = None;
+    let mut sample_members = None;
+    for ep in endpoints {
+        if !endpoints_set.insert(ep) {
+            return Err(box_err!("a duplicate PD url {}", ep));
+        }
+
+        let mut stream = match rpc_connect(ep.as_str()) {
+            Ok(stream) => stream,
+            // Ignore failed pd node.
+            Err(_) => continue,
+        };
+
+        let mut req = protocol::new_request(VALIDATE_CLUSTER_ID, pdpb::CommandType::GetPDMembers);
+        req.set_get_pd_members(pdpb::GetPDMembersRequest::new());
+        let (mid, mut resp) = match send_msg(&mut stream, VALIDATE_MSG_ID, &req) {
+            Ok((mid, resp)) => (mid, resp),
+            // Ignore failed pd node.
+            Err(_) => continue,
+        };
+
+        if mid != VALIDATE_MSG_ID {
+            return Err(box_err!("PD response msg_id mismatch, want {}, got {}",
+                                VALIDATE_MSG_ID,
+                                mid));
+        }
+
+        // Check cluster ID.
+        let cid = resp.take_header().get_cluster_id();
+        if let Some(sample) = cluster_id {
+            if sample != cid {
+                return Err(box_err!("PD response cluster_id mismatch, want {}, got {}",
+                                    sample,
+                                    cid));
+            }
+        } else {
+            cluster_id = Some(cid);
+        }
+
+        // Check all fields.
+        let mut members = resp.take_get_pd_members().take_members().into_vec();
```

Comment:
I am confused that why do we need to check this. Is it necessary? 

Or you can give some test for it and let me how this works.